### PR TITLE
Managed shmem segment remap race

### PR DESCRIPTION
**Pull request by drrtuy**

### CODE REVIEW

File: versioning/BRM/extentmap.cpp [hunk 1]
```diff
@@ -219,15 +219,20 @@ boost::mutex ExtentMapRBTreeImpl::fInstanceMutex;
 ExtentMapRBTreeImpl* ExtentMapRBTreeImpl::fInstance = nullptr;
 
 /*static*/
-ExtentMapRBTreeImpl* ExtentMapRBTreeImpl::makeExtentMapRBTreeImpl(unsigned key, off_t size, bool readOnly)
+ExtentMapRBTreeImpl* ExtentMapRBTreeImpl::makeExtentMapRBTreeImpl(unsigned key, off_t size, bool readOnly,
+                                                                  const MasterSegmentTable* emSegTable)
```
Comment (tntnatbry) [inline versioning/BRM/extentmap.cpp:223]: `emSegTable` should not be const here.

File: versioning/BRM/extentmap.cpp [hunk 2]
```diff
@@ -219,15 +219,20 @@ boost::mutex ExtentMapRBTreeImpl::fInstanceMutex;
...
 
   if (fInstance)
   {
     if (key != fInstance->fManagedShm.key())
     {
+      if (emSegTable)
+        emSegTable->getTable_upgrade(MasterSegmentTable::EMTable);
       fInstance->fManagedShm.reMapSegment();
+      if (emSegTable)
+        emSegTable->getTable_downgrade(MasterSegmentTable::EMTable);
```
Comment (tntnatbry) [inline versioning/BRM/extentmap.cpp:235]: This should be changed to:
```c++
      if (emSegTable)
      {
        emSegTable->getTable_upgrade(MasterSegmentTable::EMTable);
        emLocked = true;
      }
      fInstance->fManagedShm.reMapSegment();
      if (emSegTable)
      {
        emLocked = false;
        emSegTable->getTable_downgrade(MasterSegmentTable::EMTable);
      }
```
Obviously, the function signature would need to be changed appropriately to pass `emLocked` from the caller.

File: versioning/BRM/extentmap.h [hunk 3]
```diff
@@ -1187,6 +1187,8 @@ class ExtentMap : public Undoable
   ExtentMapRBTreeImpl* fPExtMapRBTreeImpl;
   FreeListImpl* fPFreeListImpl;
   ExtentMapIndexImpl* fPExtMapIndexImpl_;
+
+  friend class UTTester;
```
Comment (tntnatbry) [inline versioning/BRM/extentmap.h:1191]: I am not sure what this change is.

**Review by tntnatbry — CHANGES_REQUESTED** [review]

File: versioning/BRM/extentmap.cpp (hunk 1 above)
Comment (drrtuy) [reply versioning/BRM/extentmap.cpp:223]: Why?

**Comment by drrtuy** [comment]
Thank you for the review. JFYI this version was not to be reviewed. I pushed it to check the build.

File: versioning/BRM/extentmap.cpp (hunk 2 above)
Comment (drrtuy) [reply versioning/BRM/extentmap.cpp:235]: Agree

File: versioning/BRM/extentmap.cpp (hunk 1 above)
Comment (drrtuy) [reply versioning/BRM/extentmap.cpp:223]: It is a constant pointer to a MST and I don't change it so I don't see any reason to remove const here.

Comment (tntnatbry) [reply versioning/BRM/extentmap.cpp:223]: I just noticed `getTable_upgrade` and `getTable_downgrade` are const functions, so this looks good to me.

File: versioning/BRM/extentmap.h (hunk 3 above)
Comment (drrtuy) [reply versioning/BRM/extentmap.h:1191]: Doesn't matter b/c it is gone.

**Pull request closed without merging**